cache-filter/src/filter/root.rs

```diff
@@ -0,0 +1,69 @@
+use crate::configuration::FilterConfig;
+use crate::filter::http::CacheFilter;
+use log::{debug, info, warn};
+use proxy_wasm::{
+    traits::{ HttpContext, RootContext, Context },
+    types::{ ContextType, LogLevel },
+};
+
+#[no_mangle]
+pub fn _start() {
+    proxy_wasm::set_log_level(LogLevel::Trace);
+    proxy_wasm::set_root_context(|context_id| -> Box<dyn RootContext> {
+        Box::new(CacheFilterRoot {
+            context_id,
+            config: FilterConfig::default(),
+        })
+    });
+}
+
+struct CacheFilterRoot {
+    context_id: u32,
+    config: FilterConfig,
+}
+
+impl RootContext for CacheFilterRoot {
+    fn on_vm_start(&mut self, _vm_configuration_size: usize) -> bool {
+        info!("VM started");
+        true
+    }
+
+    fn on_configure(&mut self, _config_size: usize) -> bool {
+        //Check for the configuration passed by envoy.yaml
+        let configuration: Vec<u8> = match self.get_configuration() {
+            Some(c) => c,
+            None => {
+                warn!("Configuration missing. Please check the envoy.yaml file for filter configuration");
+                return false;
+            }
+        };
+
+        // Parse and store the configuration passed by envoy.yaml
+        match serde_json::from_slice::<FilterConfig>(configuration.as_ref()) {
+            Ok(config) => {
+                debug!("configuring {}: {:?}", self.context_id, config);
+                self.config = config;
+                return true;
+            }
+            Err(e) => {
+                warn!("Failed to parse envoy.yaml configuration: {:?}", e);
+                return false;
+            }
+        }
+    }
+
+    fn create_http_context(&self, context: u32) -> Option<Box<dyn HttpContext>> {
+        Some(Box::new(CacheFilter {
+            context_id: context,
+            config: self.config.clone(),
+            to_report_singleton: true,
+            cache_key: String::new(),
```

Comment:
Better to have this as a string literal in a function scope where it's needed. `String` will be allocated on the heap and since the `HTTPContext` gets created per every http request, we'll have heap allocations equal to the number of requests. String literals are in the stack and gets invalidated when it goes out of scope, it's better to have this as a string literal.

Reply:
We don't have a strict memory restriction but rather a time restriction so, no need to worry about heap/stack allocation. I chose `String` because it's a matter of who owns that memory. In the case of `String`, `CacheFilter` owns that matter and in the case of string literal, that memory lies outside the `CacheFilter`.

Reply:
Note that string literals are not in the stack, they are statically allocated when the binary is loaded and are immutable - if you need to compute the `cache_key` at runtime then you cannot use a string literal. You actually compute it at runtime with `self.cache_key = format!("{}_{}",request_data.app_id,request_data.service_id);`, so it's going to be an allocation in the heap.

That said, while there are some (convoluted) ways to avoid heap allocation it is not an issue until the functionality is complete and correct, and at that point we should still assess whether it could result in a big enough impact that merits going for it.

Reply:
Sorry my bad. Only the reference is in the stack and the memory is allocated from the heap.